pageserver/src/storage_sync.rs

```diff
@@ -796,19 +596,40 @@ pub(super) fn spawn_storage_sync_thread(
         .build()
         .context("Failed to create storage sync runtime")?;
 
+    // TODO we are able to "attach" empty tenants, but not doing it now since it might require big wait time
+    // when on-demand download gets merged, we're able to do this fast by storing timeline metadata only.
```

Comment:
I don't understand this. Why are empty tenants special? Why would they take a long time when on-demand download is merged?

Reply:
> I don't understand this

Will try to post a comment, but not sure where, since "empty tenant" idea is spread over `tenant_mgr`.

> Why are empty tenants special

Empty tenants in such collections mean that we have something on disk (a tenant dir, maybe a tenant config), that was loaded as a tenant, successfully.
Yet, its timelines were not loaded for some reason, there could be multiple
* we've just created a tenant via HTTP and had not created any timelines yet, then restarted
* some timelines were created after the tenant creation, but got corrupted somehow locally (only temp files present, etc.) and we did not load such timeline

Still, such empty tenants are needed to be loaded into memory, to avoid cases like https://github.com/neondatabase/neon/pull/2352 where we have some tenant directory locally, yet nothing in memory.

Additionally, when the remote storage is on, we could go further and fetch remote indices for every remote timeline of a tenant and try to restore things more.

> Why would they take a long time when on-demand download is merged?

That is a bad phrasing of mine: it's slow now, because we require the entire timeline to be loaded for tenant.
Plus, searching for "every index of every timeline for a tenant" in S3 is one of the slowest operations: we have to do the `gather_tenant_timelines_index_parts` that is expensive.

Yet later, when we're able to do on-demand timeline download, it would become simpler: we don't need to download anything but metadata for such timelines to "attach" them and the longest part (downloading the entire timeline) is no more present.
Later, if we're implementing tenant json indices as proposed in the atomic RFC, we won't have to do the `gather_tenant_timelines_index_parts` anymore, making it even more faster.